source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.cc

```diff
@@ -0,0 +1,327 @@
+void ClientSideWeightedRoundRobinLoadBalancer::refreshHostSource(const HostsSource& source) {
+  // insert() is used here on purpose so that we don't overwrite the index if
+  // the host source already exists. Note that host sources will never be
+  // removed, but given how uncommon this is it probably doesn't matter.
+  rr_indexes_.insert({source, seed_});
+  // If the list of hosts changes, the order of picks change. Discard the
+  // index.
+  peekahead_index_ = 0;
+
+  if (!IS_MAIN_OR_TEST_THREAD()) {
+    return;
+  }
+
+  // On the main thread ensure that all hosts have client side lb policy data.
+  addClientSideLbPolicyDataToHosts(hostSourceToHosts(source));
+}
+
+HostConstSharedPtr
+ClientSideWeightedRoundRobinLoadBalancer::chooseHost(LoadBalancerContext* context) {
+  HostConstSharedPtr host = EdfLoadBalancerBase::chooseHost(context);
+  if (context != nullptr) {
+    // Configure callbacks to receive ORCA load reports.
+    context->setOrcaLoadReportCallbacks(*this);
+  }
+  return host;
+}
+
+double ClientSideWeightedRoundRobinLoadBalancer::hostWeight(const Host& host) const {
+  ENVOY_LOG(trace, "hostWeight {} = {}", getHostAddress(&host), host.weight());
+  return host.weight();
+}
+
+HostConstSharedPtr
+ClientSideWeightedRoundRobinLoadBalancer::unweightedHostPeek(const HostVector& hosts_to_use,
+                                                             const HostsSource& source) {
+  auto i = rr_indexes_.find(source);
+  if (i == rr_indexes_.end()) {
+    return nullptr;
+  }
+  return hosts_to_use[(i->second + (peekahead_index_)++) % hosts_to_use.size()];
+}
+
+HostConstSharedPtr
+ClientSideWeightedRoundRobinLoadBalancer::unweightedHostPick(const HostVector& hosts_to_use,
+                                                             const HostsSource& source) {
+  if (peekahead_index_ > 0) {
+    --peekahead_index_;
+  }
+  // To avoid storing the RR index in the base class, we end up using a second
+  // map here with host source as the key. This means that each LB decision
+  // will require two map lookups in the unweighted case. We might consider
+  // trying to optimize this in the future.
+  ASSERT(rr_indexes_.find(source) != rr_indexes_.end());
+  return hosts_to_use[rr_indexes_[source]++ % hosts_to_use.size()];
+}
+
+absl::Status ClientSideWeightedRoundRobinLoadBalancer::onOrcaLoadReport(
+    const xds::data::orca::v3::OrcaLoadReport& orca_load_report,
+    const HostDescription& host_description) {
+  const Host* host = dynamic_cast<const Host*>(&host_description);
+  ENVOY_BUG(host != nullptr, "Unable to cast HostDescription to Host.");
+  ENVOY_LOG(trace,
+            "LoadBalancerContext::OrcaLoadReportCb "
+            "orca_load_report for {} report = {}",
+            getHostAddress(host), orca_load_report.DebugString());
+  const auto& lb_policy_data_ptr = host->lbPolicyData();
+  auto* client_side_data = dynamic_cast<ClientSideHostLbPolicyData*>(lb_policy_data_ptr.get());
```

Comment:
IMHO this is a design that will be challenging to maintain in the future.
Can the per-host LB data be stored in the LB class itself?

Reply:
I'm not going to block the PR review on this.
If desired, this can be refactored in the future (and should only be an internal code change).

Reply:
Sounds good, we can revisit this in a followup PR.